bundles/org.openhab.core.audio/src/main/java/org/openhab/core/audio/internal/AudioManagerImpl.java

```diff
@@ -122,38 +122,13 @@ public void play(@Nullable AudioStream audioStream, @Nullable String sinkId) {
     public void play(@Nullable AudioStream audioStream, @Nullable String sinkId, @Nullable PercentType volume) {
         AudioSink sink = getSink(sinkId);
         if (sink != null) {
-            PercentType oldVolume = null;
-            // set notification sound volume
-            if (volume != null) {
-                try {
-                    // get current volume
-                    oldVolume = sink.getVolume();
-                } catch (IOException e) {
-                    logger.debug("An exception occurred while getting the volume of sink '{}' : {}", sink.getId(),
-                            e.getMessage(), e);
-                }
-
-                try {
-                    sink.setVolume(volume);
-                } catch (IOException e) {
-                    logger.debug("An exception occurred while setting the volume of sink '{}' : {}", sink.getId(),
-                            e.getMessage(), e);
-                }
-            }
+            Runnable volumeRestauration = null;
             try {
-                sink.process(audioStream);
+                volumeRestauration = handleVolumeCommand(volume, sink);
+                sink.processAndComplete(audioStream).thenRun(volumeRestauration);
             } catch (UnsupportedAudioFormatException | UnsupportedAudioStreamException e) {
+                volumeRestauration.run();
                 logger.warn("Error playing '{}': {}", audioStream, e.getMessage(), e);
-            } finally {
-                if (volume != null && oldVolume != null) {
-                    // restore volume only if it was set before
-                    try {
-                        sink.setVolume(oldVolume);
-                    } catch (IOException e) {
-                        logger.debug("An exception occurred while setting the volume of sink '{}' : {}", sink.getId(),
-                                e.getMessage(), e);
-                    }
-                }
             }
         } else {
```

Comment:
You can collapse all that by using `.handle((t, v) -> {...})` instead of `.thenRun()`. If an exception occured, `t != null` and contains the exception (can be used to log the warning). There is no need for the `Runnable` variable then.

Reply:
Your proposal highlight the fact that exceptions were not handled in the  'CompletableFuture' way.
So, I rewrited it to do so. Please note that I use the method `exceptionally` instead of `handle` because Void return value makes it more adequate.

But that said, I still need the Runnable variable, because the  volume value to set back has to be computed before the processAndComplete method. Or is there something I didn't understand ?